File: client-layer/src/app/services/ride-stop.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../enviroment';

export interface StopRideRequest {
  currentLocation: string;
  stoppedAt: string;
}

export interface StopRideResponse {
  rideId: number;
  stoppedLocation: string;
  stoppedAt: string;
  recalculatedPrice: number;
  message: string;
}

@Injectable({
  providedIn: 'root'
})
export class RideStopService {
  private apiUrl = `${environment.apiUrl}/api/rides`;

  constructor(private http: HttpClient) {}
  stopRide(rideId: number, request: StopRideRequest): Observable<StopRideResponse> {
    return this.http.post<StopRideResponse>(
      `${this.apiUrl}/${rideId}/stop`,
      request
    );
  }
}

File: mobile-application/app/src/main/java/com/example/mobile_application/service/RideStopService.java

package com.example.mobile_application.service;

import com.example.mobile_application.dto.StopRideRequestDTO;
import com.example.mobile_application.dto.StopRideResponseDTO;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface RideStopService {
    @POST("api/rides/{rideId}/stop")
    Call<StopRideResponseDTO> stopRide(
            @Path("rideId") Long rideId,
            @Body StopRideRequestDTO request);
}